### DISCUSSION TITLE
Any plan to add Java SDK

### DISCUSSION THREAD

**Discussion by WangCHX** [reata/sqllineage#286] [Ideas]
pretty cool stuff. wondering if you have plan to add java SDK. 🤣

**Comment by lzzcj** [comment]
+1

**Comment by reata** [comment]
I plan to do a proof of concept later this year by using ANTLR to generate a customized sql parser, replacing the current Python library sqlparse. If this approach works, then it shall be easy to build a Java SDK since ANTLR is language-agnostic. But I don't think Java SDK is coming very soon. Let's keep this thread open to see if more people want this feature.

Moving this to discussion.

Update: POC with ANTLR: https://github.com/reata/sqllineage4j

**Comment by chundongwang** [comment]
Maybe you want to use https://github.com/frsyuki/trino_sql_parser as a reference - it uses native Trino SQL parser and returns ANTLR token 😄

**Reply by reata** [reply]
I checked the repo and it's pretty much a wrapper against Trino parser, as it says.

The downside to use a Trino native parser is that it returns AST tailored to Trino needs. To be SQL dialects agnostic, actually we'd go with a Parse Tree, which is the real output of ANTLR and at a slightly lower level than Abstract Syntax Tree.
The ANTLR tokens returned by this lib is also less useful, as it's just the token list from Statement, which loses the whole tree structure.

Feel free to check the current version of sqllineage4j, I got my inspiration directly from Trino & Spark source code. It's now working well on table level lineage. Column level lineage is WIP.

**Comment by WangCHX** [comment]
quick question, is it java11 or java17 compatible?

**Reply by reata** [reply]
It's java 11 for now.

**Comment by ma1231** [comment]
May I ask if column lineage is available now

**Reply by reata** [reply]
Due to resource constraint, I don't have plan to adding column lineage to sqllineage4j as of now.

Anyone who's interested in developing sqllineage4j, feel free to contact me. I can offer some help.
